qiime2/core/type/signature.py

```diff
@@ -289,16 +289,36 @@ def decode_parameters(self, **kwargs):
 
     def check_types(self, **kwargs):
         for name, spec in self.signature_order.items():
-            if kwargs[name] not in spec.qiime_type:
-                # A type mismatch is unacceptable unless the value is None
-                # and this parameter's default value is None.
-                if not (spec.has_default() and
-                        spec.default is None and
-                        kwargs[name] is None):
+            # A type mismatch is unacceptable unless the value is None
+            # and this parameter's default value is None.
+            if ((kwargs[name] not in spec.qiime_type) and
+                    not (spec.has_default() and spec.default is None
+                         and kwargs[name] is None)):
+
+                # handle visualization passed as parameter
+                if isinstance(kwargs[name], qiime2.sdk.Visualization):
                     raise TypeError(
-                        "Parameter %r received an argument of type %r. An "
-                        "argument of subtype %r is required." % (
-                            name, kwargs[name].type, spec.qiime_type))
+                        "Parameter %r received a Visualization (*.qzv) as an "
+                        "argument. Visualizations may not be used as inputs."
+                        % name)
+
+                # handle Artifacts
+                if isinstance(kwargs[name], qiime2.sdk.Artifact):
+                    raise TypeError(
+                        "Parameter %r requires an argument of type %r. An "
+                        "argument of type %r was passed." % (
+                            name, spec.qiime_type, kwargs[name].type))
+                # handle Metadata without displaying long Metadata repr
+                elif isinstance(kwargs[name], qiime2.Metadata):
+                    raise TypeError(
+                        "Parameter %r received a Metadata file as an argument,"
```

Comment:
Similar to my comment above --- not all `Metadata` is a file --- an Artifact API user can construct `Metadata` and provide as input without ever saving to disk (in this example, it would be called a "Metadata object"). Make sense?

Reply:
🗸

Reply:
I think we overcorrected in the wrong direction here --- for all the same reasons why we might not have "qzv files", we also can't guarantee that we have "objects" here, either.

Reply:
Handled. Thanks!